File: openCV Ball tracking/pid.py
class PID:
    def __init__(self, kp, ki, kd):
        self.kp = kp
        self.ki = ki
        self.kd = kd
        self.setpoint = 0
        self.integral = 0
        self.last_error = 0
        self.last_time = None

    def update(self, measurement, current_time):
        error = self.setpoint - measurement
        delta_time = (current_time - self.last_time) if self.last_time is not None else 0
        self.integral += error * delta_time
        derivative = (error - self.last_error) / delta_time if delta_time > 0 else 0

        # PID output
        output = (self.kp * error) + (self.ki * self.integral) + (self.kd * derivative)

        # Update last error and last time for next calculation
        self.last_error = error
        self.last_time = current_time

        return output

    def set_setpoint(self, setpoint):
        self.setpoint = setpoint
        self.integral = 0
        self.last_error = 0

File: two_servo_opencv_test/main/PID.c
#include "pid.h"

void PID_Init(PID *pid, double kp, double ki, double kd) {
    pid->kp = kp;
    pid->ki = ki;
    pid->kd = kd;
    pid->setpoint = 0.0;
    pid->integral = 0.0;
    pid->last_error = 0.0;
    pid->last_time = -1.0; // Initialize to an invalid time
}

double PID_Update(PID *pid, double measurement, double current_time) {
    double error = pid->setpoint - measurement;
    double delta_time = (pid->last_time < 0) ? 0 : (current_time - pid->last_time);
    double derivative = 0.0;

    if (delta_time > 0) {
        pid->integral += error * delta_time;
        derivative = (error - pid->last_error) / delta_time;
    }

    double output = (pid->kp * error) + (pid->ki * pid->integral) + (pid->kd * derivative);

    // Update last error and last time for next calculation
    pid->last_error = error;
    pid->last_time = current_time;

    return output;
}

void PID_SetSetpoint(PID *pid, double setpoint) {
    pid->setpoint = setpoint;
    pid->integral = 0.0;    // Reset integral
    pid->last_error = 0.0;  // Reset last error
}
